prometheus/client.go

```diff
@@ -35,11 +39,18 @@ func NewClient() (*Client, error) {
 	return &client, nil
 }
 
-// For testing / mocking
-func (in *Client) inject(api v1.API) {
+// Inject allows for replacing the API with a mock For testing
+func (in *Client) Inject(api v1.API) {
 	in.api = api
 }
 
+var injectedClient *Client
```

Comment:
I'm not sure to understand why injectClient is needed. Isn't `Inject(api)` sufficient?
I ask cause if we can avoid globals/singleton as much as possible I think it would be better.

Reply:
@lucasponce this is really just supporting SWS-163, I broke it out because it was supporting code.

Reply:
Ok I've just run the test and I understand - it's because the test invoked the whole REST handler, so it doesn't have the control on the prom client creation.

Another way to go could be to split the rest handler in 2 functions: the rest handler, which would do just minimal stuff, and an unexported function that does all the work and takes the prom client as parameter. Just the latter would be tested. This way we don't introduce globals/singleton to freeze a state. On the other hand it leaves the REST api itself untested.

Reply:
@jotak The problem is that although that works when the test code itself is calling prometheus.NewClient(), if the actual code is making the call, like in (https://github.com/swift-sunshine/swscore/blob/master/handlers/graph.go#L92), it will always get a new client object, and the test code can't inject into it.  I needed to find a way to have NewClient() return a mock to the actual code under test.  I'm open to an alternative, I just didn't see how to do it.

Reply:
@jotak, I considered that approach but thought this may be less intrusive to the real code.  But since you are uncomfortable with the singleton (I don't love it, either), I'll revisit the alternate approach and see if I hit any roadblock.  I'll set this to DNM in the interim.

Reply:
@jshaughn Actually, the question here is a bit similar to writing integration tests versus unit tests. With just unit tests we wouldn't have to freeze client's state like that. And maybe integration tests should run with a real environment setup rather than mocks.

Reply:
@jshaughn ok, anyway there's place for different views :-) it's not a blocker for me